tket/src/Circuit/Boxes.cpp

```diff
@@ -494,6 +494,298 @@ op_signature_t StabiliserAssertionBox::get_signature() const {
+      std::vector<bool> transposition_i_first = transposition_i.first;
+
+      TKET_ASSERT(transposition_i_last == transposition_j.first);
+      std::vector<bool> i_middle = transposition_i.middle;
+      std::vector<bool> j_middle = transposition_j.middle;
+
+      TKET_ASSERT(i_middle.size() == transposition_i.last.size());
+      TKET_ASSERT(j_middle.size() == transposition_i.last.size());
+      // if a transposition has already been reduced, still need to make sure we
+      // uncompute it
+      if (transposition_i_first != transposition_i_last) {
+        unsigned middle_last_distance =
+            get_hamming_distance(i_middle, transposition_i_last);
+        unsigned middle_first_distance =
+            get_hamming_distance(i_middle, transposition_i_first);
+        // this => the reduced transposition is on a good gray code between the
+        // new "first" and target
+        if (middle_first_distance < middle_last_distance &&
+            middle_first_distance > 1) {
+          transposition_i_last = transposition_i_first;
+          std::vector<bool> starting_point = transposition_i_last;
+          for (unsigned k = 0; k < i_middle.size(); k++) {
+            if (i_middle[k] == j_middle[k] &&
+                get_hamming_distance(starting_point, i_middle) > 1) {
+              starting_point[k] = i_middle[k];
+            }
+          }
+
+          cycle_transpositions[k][i].last = starting_point;
+          cycle_transpositions[k][j].first = starting_point;
+        }
+      } else {  // else in this case just find any good transposition
+        std::vector<bool> starting_point = transposition_i_last;
+        for (unsigned k = 0; k < i_middle.size(); k++) {
+          if (i_middle[k] == j_middle[k] &&
+              get_hamming_distance(starting_point, i_middle) > 1) {
+            starting_point[k] = i_middle[k];
+          }
+        }
+
+        cycle[i].last = starting_point;
+        cycle[j].first = starting_point;
+      }
+
+      ++i;
+      ++j;
+    }
+
+    return_transposition.insert(
+        return_transposition.end(), cycle.begin(), cycle.end());
+  }
+
+  return return_transposition;
+}
+
+void ToffoliBox::generate_circuit() const {
+  // This decomposition is as described on page 191, section 4.5.2 "Single
+  // qubit and CNOT gates are universal" of Nielsen & Chuang
+  std::vector<ToffoliBox::cycle_transposition_t> cycle_transpositions =
+      this->get_transpositions();
+
+  // optionally, order the transpositions and cycles to allow gate
+  // cancellation
+  cycle_transposition_t ordered_transpositions;
+  if (this->reorder_) {
+    ordered_transpositions = merge_cycles(cycle_transpositions);
+  } else {
+    for (auto &transpositions : cycle_transpositions) {
+      ordered_transpositions.insert(
+          ordered_transpositions.end(), transpositions.begin(),
+          transpositions.end());
+    }
+  }
+
+  if (ordered_transpositions.empty()) {
+    this->circ_ = std::make_shared<Circuit>();
```

Comment:
Doesn't it have to be a circuit with `n_qubits` qubits?

Reply:
Yeah I guess it does. Theres two cases where there are no transpositions: 1) The permutation is empty, 2) The provided permutation has no cycles. So I guess I'll check this->cycles_ and if its not empty count the size of basis states, else return the empty circuit.

Reply:
Done.

Reply:
Ah just saw your next comment, will update to pass an n_qubits parameter instead.

Reply:
It now uses the new n_qubits_ parameter